src/Compilers/Core/MSBuildTask/ManagedToolTask.cs

```diff
@@ -0,0 +1,97 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using Microsoft.Build.Utilities;
+
+namespace Microsoft.CodeAnalysis.BuildTasks
+{
+    public abstract class ManagedToolTask : ToolTask
+    {
+        protected abstract bool IsManagedTask { get; }
+
+        protected abstract string CommandLineArguments { get; }
+
+        protected abstract string PathToManagedTool { get; }
+
+        /// <summary>
+        /// Note: "Native" here does not neccesarily mean "native binary".
+        /// "Native" in this context means "native invocation", and running the executable directly.
+        /// </summary>
+        protected abstract string PathToNativeTool { get; }
+
+        protected sealed override string GenerateCommandLineCommands()
+        {
+            var commandLineArguments = CommandLineArguments;
+            if (IsManagedTask && IsCliHost(out string pathToDotnet))
+            {
+                var pathToTool = PathToManagedTool;
+                if (pathToTool is null)
+                {
+                    Log.LogErrorWithCodeFromResources("General_ToolFileNotFound", ToolName);
+                }
+                commandLineArguments = PrependFileToArgs(pathToTool, commandLineArguments);
+            }
+
+            return commandLineArguments;
+        }
+
+        protected sealed override string GenerateFullPathToTool()
+        {
+            if (IsManagedTask)
+            {
+                if (IsCliHost(out string pathToDotnet))
+                {
+                    return pathToDotnet;
+                }
+                else
+                {
+                    return PathToManagedTool;
+                }
+            }
+            else
+            {
+                return PathToNativeTool;
+            }
+        }
+
+        protected abstract string ToolNameWithoutExtension { get; }
+
+        protected sealed override string ToolName
+        {
+            get
+            {
+                if (CoreClrShim.IsRunningOnCoreClr)
+                {
+                    return $"{ToolNameWithoutExtension}.dll";
+                }
+                else
+                {
+                    return $"{ToolNameWithoutExtension}.exe";
+                }
+            }
+        }
+
+        private static bool IsCliHost(out string pathToDotnet)
```

Comment:
I would call this `TryGetCliHostPath`

Reply:
I'm not sure if that fully addresses the purpose of this method - if this returns false, we should be able to assume that we're not on CLI. It's more of a method that goes "are we in the CLI? And as a side effect, if we are, here's the path."

(being on coreclr without DOTNET_HOST_PATH set is a very sad situation, one that should be expected to misbehave or crash - the contract made with CLI is that it will always be set)

Reply:
I don't feel strongly about this, but my feeling is that this function is just a helper -- the logic around returning false is the responsibility of the caller. It's really up to them how severely they want to treat the result.